libs/net/nodenetworkconfigurationpolicy.py

```diff
@@ -0,0 +1,140 @@
+from copy import deepcopy
+from dataclasses import asdict, dataclass
+from typing import Any
+
+from ocp_resources.exceptions import NNCPConfigurationFailed
+from ocp_resources.node_network_configuration_policy_latest import NodeNetworkConfigurationPolicy as Nncp
+from ocp_resources.resource import ResourceEditor
+from timeout_sampler import retry
+
+WAIT_FOR_STATUS_TIMEOUT_SEC = 90
+WAIT_FOR_STATUS_INTERVAL_SEC = 5
+
+
+@dataclass
+class IP:
+    enabled: bool
+    dhcp: bool | None = None
+    auto_dns: bool | None = None
+
+
+@dataclass
+class IPv4(IP):
+    pass
+
+
+@dataclass
+class IPv6(IP):
+    autoconf: bool | None = None
+
+
+@dataclass
+class STP:
+    enabled: bool
+
+
+@dataclass
+class BridgeOptions:
+    stp: STP
+
+
+@dataclass
+class Port:
+    name: str
+
+
+@dataclass
+class Bridge:
+    options: BridgeOptions | None = None
+    port: list[Port] | None = None
+    allow_extra_patch_ports: bool | None = None
+
+
+@dataclass
+class Interface:
+    name: str
+    type: str
+    state: str
+    ipv4: IPv4 | None = None
+    ipv6: IPv6 | None = None
+    bridge: Bridge | None = None
+
+
+@dataclass
+class DesiredState:
+    """
+    Represents the desired network configuration for NMstate.
+    Following the NMstate YAML-based API specification:
+    https://nmstate.io/devel/yaml_api.html
+    """
+
+    interfaces: list[Interface]
+
+
+class NodeNetworkConfigurationPolicy(Nncp):
+    """
+    NodeNetworkConfigurationPolicy object.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        desired_state: DesiredState,
+        node_selector: dict[str, str] | None = None,
+    ):
+        """
+        Create and manage NodeNetworkConfigurationPolicy
+
+        Args:
+            name (str): Name of the NodeNetworkConfigurationPolicy object.
+            desired_state (DesiredState): Desired policy configuration - interface creation, modification or removal.
+            node_selector (dict, optional): A node selector that specifies the nodes to apply the node network
+                configuration policy to.
+        """
+        self._desired_state = desired_state
+        super().__init__(
+            name=name,
+            desired_state=asdict(desired_state, dict_factory=self._dict_normalization),
+            node_selector=node_selector,
+        )
+
+    @staticmethod
+    def _dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
```

Comment:
I'm not sure how I feel about this method... ;)
The caller is responsible for calling this class with valid keys and values, and if they don't do that - the attempt to create the resource should fail with the appropriate error (which kubernetes already does).
I don't think the resource class should cover-up for the caller sending invalid parameters.

Reply:
I do not understand your comment.
This method is used to drop `None` values and convert `_` field names to `-` in order to fit the use of `dataclass`.
basically, it is part of the mechanism to convert a structured data (`dataclass`) to a dict that is used to call the API.

It has nothing to do with the user calling anything. The user is building the structured data and if that is wrong the API will be the one who will fail it.

Reply:
https://github.com/RedHatQE/openshift-virtualization-tests/pull/218#discussion_r1983385143
Marking this conversation as resolved.